File: backend/GameApp.Application/Repositories/IEnemyRepository.cs
using GameApp.Domain.Entities;
using GameApp.Domain.ValueObjects.Enemies;

namespace GameApp.Domain.Repositories;

public interface IEnemyRepository
{
    public abstract Task<IEnumerable<Enemy>> FetchAllAsync();

    public abstract Task<Enemy?> FetchByIdAsync(Guid id);

    public abstract Task<Enemy?> FetchByName(EnemyName name);

    public abstract Task<Enemy?> DeleteAsync(Guid id);
    
    public abstract Task<Enemy?> SaveAsync(Enemy Enemy);

    public abstract Task<Enemy?> UpdateAsync(Guid id, Enemy Enemy);

    public abstract Task SeedAsync();
}

File: frontend/gameapp-frontend/src/application/repositories/ienemy-repository.ts
import { Enemy } from '../../domain/entities/enemy';
import { EnemyName } from '../../domain/value-objects/enemies/enemy-name';

export interface IEnemyRepository {
    fetchById(id: string): Promise<Enemy>;
    fetchByName(name: EnemyName): Promise<Enemy>;
    fetchAll(): Promise<Enemy[]>;
    save(enemy: Enemy): Promise<Enemy>;
    delete(id: string): Promise<Enemy>;
    update(id: string, enemy: Enemy): Promise<Enemy>;
}
